apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -107,41 +111,91 @@
 background_dict = {}
 foreground_dict = {}
 
-# Calculate and print similarity scores
-for emoji_name in emoji_names_list:
-    palette_scores = []
-    for palette_word in palette_model_descriptive_names:
-        id_token = nlp(emoji_name)
-        palette_token = nlp(palette_word)
-        similarity_score = id_token.similarity(palette_token)
-        palette_scores.append(round(similarity_score, 2))
+# Define caches to store ada's raw embedding outputs to reduce query costs
+# These caches are stored as pickle files; python's native way to serialize data
+embedding_paths = ['apps/static/dance/ai/model/input_embeddings.pkl',
+'apps/static/dance/ai/model/palette_embeddings.pkl',
+'apps/static/dance/ai/model/background_embeddings.pkl',
+'apps/static/dance/ai/model/foreground_embeddings.pkl']
+input_path = embedding_paths[0]
+palette_path = embedding_paths[1]
+background_path = embedding_paths[2]
+foreground_path = embedding_paths[3]
+
+def load_embedding_cache(path):
+    # load the cache if it exists, and save a copy to disk
+    try:
+        embedding_cache = pd.read_pickle(path)
+    except FileNotFoundError:
+        embedding_cache = {}
+    with open(path, "wb") as embedding_cache_file:
+        pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache
+
+caches = [load_embedding_cache(path) for path in embedding_paths]
+input_cache = caches[0]
+palette_cache = caches[1]
+background_cache = caches[2]
+foreground_cache = caches[3]
+
+# Define a function to retrieve embeddings from the cache if present, and otherwise request via the API
+def retrieve_embedding(string: str,
+    cache_path: str,
+    embedding_cache,
+    model: str = EMBEDDING_MODEL,
+) -> list:
+    """Return embedding of given string, using a cache to avoid recomputing."""
+    if (string, model) not in embedding_cache.keys():
+        embedding_cache[(string, model)] = get_embedding(string, model)
+        with open(cache_path, "wb") as embedding_cache_file:
```

Comment:
I think it'd be more clear if you moved the logic to get embeddings from OpenAI to the `load_embedding_cache` function (and renamed to something like `load_embeddings`, I could imagine hitting OpenAI in your `except FileNotFoundError:` statement?), then the logic in here just retrieves embeddings (rather than retrieving and maybe sometimes updating embeddings if we don't have them)

Reply:
Does it seem like you'd have to iterate through the strings more than once then? The first time to figure out which embeddings are missing and add them to cache file, and then the second time to retrieve the embedding?

Reply:
Aren't we already iterating through them twice (looks like we iterate through once on line 137, then again on 167)?

Reply:
Yes, iterating through all strings to call `retrieve_embeddings` and when we call on `calculate_similarity_score` comparing each emoji with the list of outputs (palettes, background effects, foreground effects).

I moved the functions to the `HoC2023AiHelperFunctions` file and did a bunch of refactoring. To me, it seems to make sense to keep `load_embeddings_cache` which loads/opens a pickle file and `retrieve_embedding` which takes a string and gets the embedding for the string and calls on the Open API endpoint if it doesn't. Lmk what you think after seeing things updated. Thanks!

Reply:
We discussed offline and leaving functions as is. But re-wrote the snippet which calls `retrieve_embeddings` based on @bencodeorg 's feedback.